File: app/src/main/java/com/example/learnwordstrainer/ui/repetition/RepetitionViewModel.java
package com.example.learnwordstrainer.ui.repetition;

import android.app.Application;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.annotation.NonNull;

import com.example.learnwordstrainer.domain.model.Word;
import com.example.learnwordstrainer.data.repository.WordRepository;

import java.util.List;
import java.util.Random;

public class RepetitionViewModel extends AndroidViewModel {
    //private final WordRepository wordRepository;

    private final MutableLiveData<Word> currentWord = new MutableLiveData<>();
    private final MutableLiveData<List<String>> answerOptions = new MutableLiveData<>();
    private final MutableLiveData<Integer> correctAnswerIndex = new MutableLiveData<>();
    private final MutableLiveData<Integer> correctAnswerCount = new MutableLiveData<>();
    private final MutableLiveData<Integer> wrongAnswerCount = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isCorrectAnswer = new MutableLiveData<>();
    private final MutableLiveData<Boolean> isRoundCompleted = new MutableLiveData<>();
    private final MutableLiveData<Boolean> showResultFooter = new MutableLiveData<>(false);

    public RepetitionViewModel(@NonNull Application application) {
        super(application);
        //wordRepository = new WordRepository(application);
    }

    public LiveData<Word> getCurrentWord() {
        return currentWord;
    }

    public LiveData<List<String>> getAnswerOptions() {
        return answerOptions;
    }

    public LiveData<Integer> getCorrectAnswerIndex() {
        return correctAnswerIndex;
    }

    public LiveData<Integer> getCorrectAnswerCount() {
        return correctAnswerCount;
    }

    public LiveData<Integer> getWrongAnswerCount() {
        return wrongAnswerCount;
    }

    public LiveData<Boolean> getIsCorrectAnswer() {
        return isCorrectAnswer;
    }

    public LiveData<Boolean> getIsRoundCompleted() {
        return isRoundCompleted;
    }

    public LiveData<Boolean> getShowResultFooter() {
        return showResultFooter;
    }

//    public void startNewRound() {
//        isRoundCompleted.setValue(false);
//        loadRandomWord();
//        loadAnswerOptions();
//    }
//
//    private void loadRandomWord() {
//        //Word randomWord = wordRepository.getRandomWord();
//        //currentWord.setValue(randomWord);
//        //correctAnswerCount.setValue(randomWord.getCorrectAnswerCount());
//        //setValue(randomWord.getWrongAnswerCount());
//    }
//
//    private void loadAnswerOptions() {
//        Word word = currentWord.getValue();
//        if (word != null) {
//            //List<String> options = wordRepository.getRandomTranslations(word.getId());
//            int index = new Random().nextInt(4);
//            //options.add(index, word.getTranslation());
//
//            //answerOptions.setValue(options);
//            correctAnswerIndex.setValue(index);
//        }
//    }
//
//    public void checkAnswer(int selectedIndex) {
//        Word word = currentWord.getValue();
//        if (word == null) return;
//
//        int correct = word.getCorrectAnswerCount();
//        int wrong = word.getWrongAnswerCount();
//
//        //boolean isCorrect = selectedIndex == correctAnswerIndex.getValue();
//        //if (isCorrect) {
//            correct++;
//            correctAnswerCount.setValue(correct);
//        //} else {
//        //    wrong++;
//            wrongAnswerCount.setValue(wrong);
//        }
//
//        wordRepository.updateScore(word.getId(), correct, wrong);
//        isCorrectAnswer.setValue(isCorrect);
//        isRoundCompleted.setValue(true);
//        showResultFooter.setValue(true);
//    }
//
//    public void hideResultFooter() {
//        showResultFooter.setValue(false);
//    }
//
//    public String getWordForSpeech() {
//        Word word = currentWord.getValue();
//        return word != null ? word.getEnglishWord() : "";
//    }
}

File: app/src/main/java/com/example/learnwordstrainer/ui/repetition/RepetitionViewModel.kt
package com.example.learnwordstrainer.ui.repetition

import android.content.Context
import androidx.appcompat.app.AppCompatDelegate
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.example.learnwordstrainer.R
import com.example.learnwordstrainer.di.IoDispatcher
import com.example.learnwordstrainer.domain.model.DailyStatistic
import com.example.learnwordstrainer.domain.repository.ThemeRepository
import com.example.learnwordstrainer.domain.usecase.GetRepetitionWordUseCase
import com.example.learnwordstrainer.domain.usecase.GetTodayStatsUseCase
import com.example.learnwordstrainer.domain.usecase.StatType
import com.example.learnwordstrainer.domain.usecase.UpdateDailyStatsUseCase
import com.example.learnwordstrainer.domain.usecase.UpdateWordScoreUseCase
import com.example.learnwordstrainer.ui.repetition.state.RepetitionUiState
import com.example.learnwordstrainer.utils.TTSManager
import dagger.hilt.android.lifecycle.HiltViewModel
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CoroutineDispatcher
import kotlinx.coroutines.flow.*
import kotlinx.coroutines.launch
import java.util.concurrent.CancellationException
import javax.inject.Inject

@HiltViewModel
class RepetitionViewModel @Inject constructor(
    @ApplicationContext private val context: Context,
    themeRepository: ThemeRepository,
    private val getRepetitionWordUseCase: GetRepetitionWordUseCase,
    private val updateWordScoreUseCase: UpdateWordScoreUseCase,
    private val updateDailyStatsUseCase: UpdateDailyStatsUseCase,
    private val ttsManager: TTSManager,
    getTodayStatsUseCase: GetTodayStatsUseCase,
    @IoDispatcher private val ioDispatcher: CoroutineDispatcher
) : ViewModel() {

    val currentTheme: StateFlow<Int> = themeRepository.themeMode
        .stateIn(
            scope = viewModelScope,
            started = SharingStarted.WhileSubscribed(5000),
            initialValue = AppCompatDelegate.MODE_NIGHT_FOLLOW_SYSTEM
        )

    private val dailyStatisticFlow: Flow<DailyStatistic?> = getTodayStatsUseCase()

    private val _uiState = MutableStateFlow<RepetitionUiState>(RepetitionUiState.Loading)
    val uiState = _uiState.asStateFlow()

    private var isProcessingAnswer = false

    init {
        loadNextWord()
    }

    override fun onCleared() {
        super.onCleared()
        ttsManager.shutdown()
    }

    fun onEvent(event: RepetitionEvent) {
        when (event) {
            is RepetitionEvent.OnAnswerSelected -> {
                if (!isProcessingAnswer) {
                    handleAnswerSelection(event.index)
                }
            }
            is RepetitionEvent.OnNextWordClicked -> loadNextWord()
            is RepetitionEvent.OnListenClicked -> handleListen()
        }
    }

    private fun handleListen() {
        val currentState = _uiState.value
        if (currentState is RepetitionUiState.Content) {
            ttsManager.speak(currentState.word.sourceWord)
        }
    }

    private fun handleAnswerSelection(selectedIndex: Int) {
        val currentState = _uiState.value
        if (currentState is RepetitionUiState.Content &&
            currentState.selectedAnswerIndex == null &&
            selectedIndex < currentState.answerOptions.size) {

            isProcessingAnswer = true
            val selectedOption = currentState.answerOptions[selectedIndex]
            val isCorrect = selectedOption == currentState.word.translation

            viewModelScope.launch(ioDispatcher) {
                try {
                    updateWordScoreUseCase(
                        wordId = currentState.word.id,
                        currentCorrectCount = currentState.word.correctAnswerCount,
                        currentWrongCount = currentState.word.wrongAnswerCount,
                        wasAnswerCorrect = isCorrect
                    )
                    val statType = if (isCorrect) StatType.CORRECT_ANSWER else StatType.WRONG_ANSWER
                    updateDailyStatsUseCase(statType)

                    _uiState.update {
                        (it as RepetitionUiState.Content).copy(
                            selectedAnswerIndex = selectedIndex,
                            isAnswerCorrect = isCorrect
                        )
                    }
                } catch (e: Exception) {
                    if (e is CancellationException) throw e
                    _uiState.value = RepetitionUiState.Error(context.getString(R.string.error_failed_to_save_answer))
                } finally {
                    isProcessingAnswer = false
                }
            }
        }
    }

    private fun loadNextWord() {
        viewModelScope.launch(ioDispatcher) {
            _uiState.value = RepetitionUiState.Loading
            try {
                // Виклик юзкейса залишається тим самим, але тепер він враховує мову
                val repetitionData = getRepetitionWordUseCase()
                if (repetitionData != null) {
                    val currentStats = dailyStatisticFlow.firstOrNull()
                    _uiState.value = RepetitionUiState.Content(
                        word = repetitionData.word,
                        answerOptions = repetitionData.options,
                        correctCount = repetitionData.word.correctAnswerCount,
                        wrongCount = repetitionData.word.wrongAnswerCount,
                        dailyStats = currentStats
                    )
                } else {
                    // ЗМІНА: Показуємо більш конкретну помилку, якщо для обраної мови немає слів
                    _uiState.value = RepetitionUiState.Error(context.getString(R.string.error_no_words_for_language))
                }
            } catch (e: Exception) {
                if (e is CancellationException) throw e
                _uiState.value = RepetitionUiState.Error(context.getString(R.string.error_failed_to_load_word))
            }
        }
    }
}

sealed interface RepetitionEvent {
    data class OnAnswerSelected(val index: Int) : RepetitionEvent
    object OnNextWordClicked : RepetitionEvent
    object OnListenClicked : RepetitionEvent
}